### PR TITLE
[Feature] Support Mamba-Codestral-7B-v0.1 on Ascend 910b

### PR DESCRIPTION
**Pull request by bingli-2026**
### What this PR does

This PR adds and validates `mistralai/Mamba-Codestral-7B-v0.1` on Ascend 910b.

Main changes:

- Add pure `MambaSpec` KV cache allocation support in Ascend 910b model runner.
- Preserve `is_prefilling` in Ascend common attention metadata for Mamba execution.
- Add compatibility patches for Mamba SSD Triton wrappers on `triton-ascend==3.2.0`.
- Allow Ascend Mamba causal conv wrappers to ignore upstream extra kwargs.
- Add a fallback slot mapping path for non-Triton placeholder kernel cases.
- Add e2e accuracy config and A2 accuracy group entry for Mamba-Codestral.
- Add English deployment documentation.
- Add the model to `.github/workflows/misc/model_dataset_list.json`.

### Validation

Environment:

- Hardware: Ascend 910B2C, 64 GB HBM, single card
- CANN: 8.5.1 (driver: 25.5.1)
- Python: 3.11
- torch: 2.9.0+cpu
- torch-npu: 2.9.0
- vLLM: 0.19.0
- Triton backend: `triton-ascend==3.2.0`

Smoke regression:

- `GET /health`: HTTP 200
- `GET /v1/models`: HTTP 200
- `POST /v1/completions` with `def fibonacci(n):`: HTTP 200
- `POST /v1/completions` with SQL prompt: HTTP 200
- `POST /v1/chat/completions`: HTTP 400 because this tokenizer has no chat template. The documented path uses `/v1/completions`.

Accuracy command:

```bash
python -m pytest tests/e2e/models/test_lm_eval_correctness.py \
  --config /data/accuracy_logs/Mamba-Codestral-7B-v0.1.local.yaml \
  --tp-size 1 \
  --report-dir /data/accuracy_logs/reports \
  -s
```

Accuracy result:

```text
gsm8k | exact_match,strict-match: ground_truth=0.1243 | measured=0.1243 | success=✅
gsm8k | exact_match,flexible-extract: ground_truth=0.1259 | measured=0.1259 | success=✅
1 passed, 5 warnings in 9910.75s (2:45:10)
pytest_status=0
```

Log paths from the Ascend validation environment:

- Smoke regression: `/data/regression_mamba_codestral_20260411.log`
- Accuracy pass log: `/data/accuracy_logs/mamba_codestral_accuracy_local_20260411_141838.log`
- Accuracy report: `/data/accuracy_logs/reports/Mamba-Codestral-7B-v0.1.md`

### Known non-blocking warnings

The run logs include:

```text
Failed to import Triton kernels. Error: No module named 'triton.language.target_info'
```

This is a compatibility warning from some vLLM Triton kernel import paths with `triton-ascend==3.2.0`. It did not block Mamba-Codestral serving or the full GSM8K e2e accuracy run.

The logs also contain repeated `Please DO NOT tune args ...` warnings from `triton-ascend` kernel compilation/autotune paths. They are noisy but non-blocking; the full accuracy run completed successfully.

Related issue: #7337

- vLLM version: v0.20.2
- vLLM main: https://github.com/vllm-project/vllm/commit/0d4d334eaa583b9c09aa4eb7538c22db99fd84b3

### CODE REVIEW

**Comment by bingli-2026** [comment]
### Ascend validation evidence

Local Ascend validation was completed on a single Ascend 910B2C card.

Environment:

```text
Hardware: Ascend 910B2C, 64 GB HBM, single card
CANN: 8.5.1
Driver: 25.5.1
Python: 3.11
torch: 2.9.0+cpu
torch-npu: 2.9.0
vLLM: 0.19.0
Triton backend: triton-ascend==3.2.0
```

Smoke regression log path in the validation environment:

```text
/data/regression_mamba_codestral_20260411.log
```

Smoke result summary:

```text
HAS_TRITON True
triton 3.2.0 /data/venv/lib/python3.11/site-packages/triton/__init__.py
GET /health -> HTTP 200
GET /v1/models -> HTTP 200
POST /v1/completions fibonacci -> HTTP 200
POST /v1/completions SQL -> HTTP 200
POST /v1/chat/completions -> HTTP 400
```

`/v1/chat/completions` returns 400 because the tokenizer does not define a chat template. This is not a model execution failure; the deployment document uses `/v1/completions`.

Accuracy pass log path in the validation environment:

```text
/data/accuracy_logs/mamba_codestral_accuracy_local_20260411_141838.log
```

Accuracy report path:

```text
/data/accuracy_logs/reports/Mamba-Codestral-7B-v0.1.md
```

Accuracy command:

```bash
python -m pytest tests/e2e/models/test_lm_eval_correctness.py \
  --config /data/accuracy_logs/Mamba-Codestral-7B-v0.1.local.yaml \
  --tp-size 1 \
  --report-dir /data/accuracy_logs/reports \
  -s
```

Key accuracy log excerpt:

```text
gsm8k | exact_match,strict-match: ground_truth=0.1243 | measured=0.1243 | success=✅
gsm8k | exact_match,flexible-extract: ground_truth=0.1259 | measured=0.1259 | success=✅
1 passed, 5 warnings in 9910.75s (2:45:10)
pytest_status=0
```

Known non-blocking warnings observed in logs:

```text
Failed to import Triton kernels. Error: No module named 'triton.language.target_info'
```

This warning comes from vLLM Triton import paths that are not compatible with `triton-ascend==3.2.0`. It did not block serving or the full GSM8K e2e accuracy run.

The logs also contain repeated `Please DO NOT tune args ...` warnings from `triton-ascend` kernel compilation/autotune paths. They are noisy but non-blocking; the full accuracy run completed successfully.

**Comment by bingli-2026** [comment]
### SKILL / Prompts used

Task objective:

- Adapt and validate `mistralai/Mamba-Codestral-7B-v0.1` on vllm-ascend Ascend 910b.
- Prefer the official compatibility matrix and use `triton-ascend==3.2.0`.
- Use fallback only with evidence.
- Preserve logs for PR submission.

Debugging prompts and workflow:

- Investigate why `Mamba-Codestral-7B-v0.1` fails to start on Ascend 910b.
- Explain what Triton is, why it can be downgraded/replaced, and what role Triton kernels have in this model path.
- Check whether `triton-ascend==3.2.0` can replace upstream Triton on Ascend 910B.
- Run smoke serving validation and full e2e accuracy validation.
- Preserve logs and summarize non-blocking warnings separately from real failures.

Key implementation decisions:

- Add pure `MambaSpec` KV cache allocation because this model has no attention layer names matching the existing allocation path.
- Patch Mamba SSD wrapper calls in vllm-ascend rather than modifying upstream vLLM site-packages.
- Preserve `is_prefilling` in Ascend common attention metadata to match upstream Mamba backend expectations.
- Use `apply_chat_template: False` and `fewshot_as_multiturn: False` because this tokenizer has no chat template.

**Comment by bingli-2026** [comment]
### PR template metadata

- vLLM version used for validation: `0.19.0`
- vLLM source install in validation environment: `/data/venv`, commit reported as `unknown` by the e2e report
- vLLM Ascend validation source: `/data/vllm-ascend`, code-equivalent to this PR for the runtime patch files by SHA256 comparison
- Validation hardware: single Ascend 910B2C, CANN `8.5.1`, driver `25.5.1`

Note: the validation environment was an editable source install and the e2e report could not resolve git commit metadata because the runtime tree was not a git checkout.

**Comment by bingli-2026** [comment]
### Follow-up fixes for CI/review feedback

Pushed commit `92e43870` to address the current PR failures and automated review feedback.

Changes in this follow-up:

- Added `docs/source/tutorials/models/Mamba-Codestral-7B-v0.1.md` to the model tutorial toctree. This was the likely cause of the ReadTheDocs failure because `.readthedocs.yaml` uses `fail_on_warning: true`.
- Updated the Mamba SSD state-passing patch to use `torch.accelerator.device_index(states.device)` instead of `states.device.index`.
- Kept `_state_passing_fwd` API-compatible by dispatching based on the upstream helper signature, preserving the `last_chunk_indices` path and adding a fallback for the older `cu_chunk_seqlens` / `seq_idx` call shape.
- Narrowed slot mapping fallback initialization to `slot_mapping_np[:num_tokens]`.

Additional local validation after the fix:

```text
python -m py_compile vllm_ascend/patch/worker/patch_mamba_ssd.py vllm_ascend/worker/block_table.py
# passed

git diff --check
# passed

.venv-pr/bin/python -m sphinx -b html docs/source /tmp/vllm-ascend-docs -W
# build succeeded
```

I also attempted to run the related unit tests in a local isolated venv, but stopped at the expected runtime dependency boundary (`torch` missing). The full Ascend e2e accuracy evidence remains in the earlier validation comment.

File: vllm_ascend/patch/worker/patch_mamba_ssd.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        (initial_states.stride(0), initial_states.stride(1), initial_states.stride(2))
+        if initial_states is not None
+        else (0, 0, 0)
+    )
+    if initial_states_ptr is None:
+        # Match _chunk_scan_fwd: keep HAS_INITSTATES false while giving Ascend a typed ptr.
+        initial_states_ptr = states.new_empty((1, 1, 1))
+
+    grid = lambda META: (_state_passing.triton.cdiv(dim, META["BLOCK_SIZE"]), batch, nheads)
+    with torch.accelerator.device_index(states.device.index):
```
[the remark this answers is not in the document]
Comment (bingli-2026) [reply vllm_ascend/patch/worker/patch_mamba_ssd.py]: Addressed in `92e43870` by passing the full `torch.device` object to `torch.accelerator.device_index(states.device)`.

File: vllm_ascend/patch/worker/patch_mamba_ssd.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        HAS_INITSTATES=initial_states is not None,
+    )
+    return
+
+
+def _state_passing_fwd(
+    states,
+    dA_cumsum,
+    last_chunk_indices,
+    initial_states=None,
```
[the remark this answers is not in the document]
Comment (bingli-2026) [reply vllm_ascend/patch/worker/patch_mamba_ssd.py:167]: Addressed in `92e43870`. The patched `_state_passing_fwd` now accepts `*args`/`**kwargs` and dispatches based on the upstream helper signature, preserving the `last_chunk_indices` path while keeping a compatibility path for `cu_chunk_seqlens` / `seq_idx`.

**Comment by wjunLu** [comment]
Do you really sure CANN is `25.5.1`?
I think you may mean driver version is `25.5.1`

**Comment by bingli-2026** [comment]
Thanks for the catch.

You are right: `25.5.1` is the driver version. I have corrected the PR evidence/metadata comments to:

- CANN: `8.5.1`
- Driver: `25.5.1`

No code logic change in this update, only environment description correction.

**Comment by wjunLu** [comment]
What is Ascend V1? I never heard of that
